Sources/GRPCCore/Call/Server/ServerInterceptorOperation.swift

```diff
@@ -0,0 +1,107 @@
+///  ``ClientInterceptorOperation`` for the client-side version of this type.
+public struct ServerInterceptorOperation: Sendable {
+  internal enum Wrapped: Sendable {
+    case allServices(interceptor: any ServerInterceptor)
+    case serviceSpecific(interceptor: any ServerInterceptor, services: [String])
+    case methodSpecific(interceptor: any ServerInterceptor, methods: [MethodDescriptor])
+  }
+
+  /// An operation specifying an interceptor that applies to all RPCs across all services will be registered with this server.
+  /// - Parameter interceptor: The interceptor to register with the server.
+  /// - Returns: A ``ServerInterceptorOperation``.
+  public static func applyToAllServices(
+    _ interceptor: any ServerInterceptor
+  ) -> Self {
+    Self(wrapped: .allServices(interceptor: interceptor))
+  }
+
+  /// An operation specifying an interceptor that will be applied only to RPCs directed to the specified services.
+  /// - Parameters:
+  ///   - interceptor: The interceptor to register with the server.
+  ///   - services: The list of service names for which this interceptor should intercept RPCs.
+  /// - Returns: A ``ServerInterceptorOperation``.
+  public static func apply(
+    _ interceptor: any ServerInterceptor,
+    onlyToServices services: [String]
+  ) -> Self {
+    Self(
+      wrapped: .serviceSpecific(
+        interceptor: interceptor,
+        services: services
+      )
+    )
+  }
+
+  /// An operation specifying an interceptor that will be applied only to RPCs directed to the specified service methods.
+  /// - Parameters:
+  ///   - interceptor: The interceptor to register with the server.
+  ///   - services: The list of method descriptors for which this interceptor should intercept RPCs.
+  /// - Returns: A ``ServerInterceptorOperation``.
+  public static func apply(
+    _ interceptor: any ServerInterceptor,
+    onlyToMethods methods: [MethodDescriptor]
+  ) -> Self {
+    Self(
+      wrapped: .methodSpecific(
+        interceptor: interceptor,
+        methods: methods
+      )
+    )
+  }
+
+  private let wrapped: Wrapped
+
+  private init(wrapped: Wrapped) {
+    self.wrapped = wrapped
+  }
+
+  /// Get the ``ServerInterceptor`` associated with this ``ServerInterceptorOperation``.
+  public var interceptor: any ServerInterceptor {
+    switch self.wrapped {
+    case .allServices(let interceptor):
+      return interceptor
+    case .serviceSpecific(let interceptor, _):
+      return interceptor
+    case .methodSpecific(let interceptor, _):
+      return interceptor
+    }
+  }
+
+  /// Returns whether this ``ServerInterceptorOperation`` applies to the given `descriptor`.
+  /// - Parameter descriptor: A ``MethodDescriptor`` for which to test whether this interceptor applies.
+  /// - Returns: `true` if this interceptor applies to the given `descriptor`, or `false` otherwise.
+  public func applies(to descriptor: MethodDescriptor) -> Bool {
```

Comment:
Does this need to be public?

Reply:
Only if we want to keep `RPCRouter/registerInterceptors` inlinable.

Reply:
Okay, in that case we should prefix it with an underscore